service/history/workflowTaskHandler.go

```diff
@@ -229,19 +242,87 @@ func (handler *workflowTaskHandlerImpl) handleCommandScheduleActivity(
 	)
 	if err != nil || failWorkflow {
 		handler.stopProcessing = true
-		return err
+		return nil, err
 	}
 
 	enums.SetDefaultTaskQueueKind(&attr.GetTaskQueue().Kind)
 
-	_, _, err = handler.mutableState.AddActivityTaskScheduledEvent(handler.workflowTaskCompletedID, attr)
+	localDispatchActivity := false
+	namespace := handler.mutableState.GetNamespaceEntry().Name().String()
+	if attr.RequestStart &&
+		targetNamespaceID == namespaceID && // TODO: shall we return error, other check?
+		handler.config.EnableActivityLocalDispatch(namespace) {
+		localDispatchActivity = true
+	}
+
+	event, ai, err := handler.mutableState.AddActivityTaskScheduledEvent(
+		handler.workflowTaskCompletedID,
+		attr,
+		localDispatchActivity,
+	)
 	if err != nil {
 		if _, ok := err.(*serviceerror.InvalidArgument); ok {
-			return handler.failCommand(enumspb.WORKFLOW_TASK_FAILED_CAUSE_SCHEDULE_ACTIVITY_DUPLICATE_ID, err)
+			return nil, handler.failCommand(enumspb.WORKFLOW_TASK_FAILED_CAUSE_SCHEDULE_ACTIVITY_DUPLICATE_ID, err)
 		}
-		return err
+		return nil, err
 	}
-	return nil
+
+	if !localDispatchActivity {
+		return nil, nil
+	}
+
+	if _, err := handler.mutableState.AddActivityTaskStartedEvent(
+		ai,
+		event.GetEventId(),
+		uuid.New(),
+		handler.identity,
+	); err != nil {
+		return nil, err
+	}
+	return func(resp *historyservice.RespondWorkflowTaskCompletedResponse) error {
+		runID := handler.mutableState.GetExecutionState().RunId
+		attr := event.GetActivityTaskScheduledEventAttributes()
+
+		taskToken := &tokenspb.Task{
+			NamespaceId:     namespaceID.String(),
+			WorkflowId:      executionInfo.WorkflowId,
+			RunId:           runID,
+			ScheduleId:      event.EventId,
+			ScheduleAttempt: ai.Attempt,
+			ActivityId:      attr.ActivityId,
+			ActivityType:    attr.ActivityType.GetName(),
+		}
+		serializedToken, err := handler.tokenSerializer.Serialize(taskToken)
+		if err != nil {
+			return err
+		}
+		response := &workflowservice.PollActivityTaskQueueResponse{
+			ActivityId:   attr.ActivityId,
+			ActivityType: attr.ActivityType,
+			Header:       attr.Header,
+			Input:        attr.Input,
+			WorkflowExecution: &commonpb.WorkflowExecution{
+				WorkflowId: executionInfo.WorkflowId,
+				RunId:      runID,
+			},
+			CurrentAttemptScheduledTime: ai.ScheduledTime,
+			ScheduledTime:               event.EventTime,
+			ScheduleToCloseTimeout:      attr.ScheduleToCloseTimeout,
+			StartedTime:                 ai.StartedTime,
+			StartToCloseTimeout:         attr.StartToCloseTimeout,
+			HeartbeatTimeout:            attr.HeartbeatTimeout,
+			TaskToken:                   serializedToken,
+			Attempt:                     ai.Attempt,
+			HeartbeatDetails:            ai.LastHeartbeatDetails,
+			WorkflowType:                handler.mutableState.GetWorkflowType(),
+			WorkflowNamespace:           namespace,
+
+			// RetryPolicy: TODO: do we need this field?
+			// not specified by matching or frontend as well
```

Comment:
it seems this is not used by sdk (checked sdk-go). I think we should not have this field.

Reply:
Removed the TODO. I will double check with SDK side before removing it from proto def.